### PR TITLE
Feature/jwt auth

### PR DESCRIPTION
**Pull request by vc-bhanudatsinh**
Code push of jwt auth

### CODE REVIEW

File: JWT_Auth/src/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import bcrypt from "bcryptjs";
+import jwt from "jsonwebtoken";
+import dotenv from "dotenv";
+
+import * as helper from "../utils/helper.js";
+
+dotenv.config();
```
Comment (karansoni-vc) [inline JWT_Auth/src/controllers/authController.js]: No need to write dotenv.config() in each file for env variables, just need to add at entry point

File: JWT_Auth/src/utils/helper.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    if (!fs.existsSync(dbFilePath)) {
+      // create a file for user data
+      await fsPromise.writeFile(dbFilePath, "[]", { encoding: "utf-8" });
+    }
+
+    // return users data from file.
+    let fileData = await fsPromise.readFile(dbFilePath, {
+      encoding: "utf-8",
+    });
+    return (fileData = JSON.parse(fileData));
```
Comment (karansoni-vc) [inline JWT_Auth/src/utils/helper.js]: why reassigning parsed data to fileData ? remove it and make variable const

File: JWT_Auth/src/utils/helper.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import * as fsPromise from "fs/promises";
+import dotenv from "dotenv";
+
+dotenv.config();
+
+const dbFilePath = process.env.DB_PATH;
+
+export const handleResponseSend = async (res, message, code) => {
+  res.statusCode = code;
+  return res.end(JSON.stringify({ message }));
```
Comment (karansoni-vc) [inline JWT_Auth/src/utils/helper.js]: why stringify the response obj ? remove it

**Review by karansoni-vc — CHANGES_REQUESTED** [review]

File: JWT_Auth/src/utils/helper.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+export const getFileData = async (res) => {
+  try {
+    // check file exists or not
+    if (!fs.existsSync(dbFilePath)) {
+      // create a file for user data
+      await fsPromise.writeFile(dbFilePath, "[]", { encoding: "utf-8" });
+    }
+
+    // return users data from file.
+    let fileData = await fsPromise.readFile(dbFilePath, {
```
Comment (karansoni-vc) [inline JWT_Auth/src/utils/helper.js]: this is correct but you can also read file directly by importing

File: JWT_Auth/src/controllers/authController.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+export const signupUser = async (req, res, next) => {
+  try {
+    const { name, email, password } = req.body;
+    const dbData = await helper.getFileData();
+    const isEmailExist = dbData.find((user) => user.email === email);
+    if (isEmailExist)
+      return res.status(200).send({ message: "User already exists" });
+    const hashPassword = await bcrypt.hash(password, 10);
+    dbData.push({ name, email, password: hashPassword });
+    await helper.writeDataInDb(dbData);
```
Comment (karansoni-vc) [inline JWT_Auth/src/controllers/authController.js]: not passing res ?

**Review by karansoni-vc — CHANGES_REQUESTED** [review]

File: JWT_Auth/src/controllers/authController.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import * as helper from "../utils/helper.js";
+
+dotenv.config();
+
+export const loginUser = async (req, res, next) => {
+  try {
+    const { email, password } = req.body;
+    const dbData = await helper.getFileData();
+    const user = dbData.find((user) => user.email === email);
+    if (!user) return res.status(200).send({ message: "Email not found" });
```
Comment (karansoni-vc) [inline JWT_Auth/src/controllers/authController.js]: incorrect statuscode

File: JWT_Auth/src/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+export const getUser = async (req, res, next) => {
+  try {
+    let dbData = await helper.getFileData();
+
+    dbData = dbData.map((user) => {
+      delete user["password"];
+      return user;
+    });
+    return res.status(200).send({ data: dbData });
```
Comment (karansoni-vc) [inline JWT_Auth/src/controllers/userController.js]: make a common success/error response pattern

File: JWT_Auth/src/utils/helper.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      encoding: "utf-8",
+    });
+    return (fileData = JSON.parse(fileData));
+  } catch (error) {
+    throw error;
```
Comment (karansoni-vc) [inline JWT_Auth/src/utils/helper.js]: return proper error

**Review by karansoni-vc — DISMISSED** [review]

File: JWT_Auth/src/controllers/authController.js (hunk 1 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/controllers/authController.js]: ok

File: JWT_Auth/src/utils/helper.js (hunk 2 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/utils/helper.js]: ok

File: JWT_Auth/src/utils/helper.js (hunk 3 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/utils/helper.js]: ok

File: JWT_Auth/src/utils/helper.js (hunk 4 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/utils/helper.js]: ok

File: JWT_Auth/src/controllers/authController.js (hunk 5 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/controllers/authController.js]: ok

File: JWT_Auth/src/controllers/authController.js (hunk 6 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/controllers/authController.js]: ok

File: JWT_Auth/src/controllers/userController.js (hunk 7 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/controllers/userController.js]: ok

File: JWT_Auth/src/utils/helper.js (hunk 8 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/src/utils/helper.js]: ok

File: JWT_Auth/package.json [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "bcryptjs": "^2.4.3",
+    "dotenv": "^16.0.3",
+    "express": "^4.18.2",
+    "jsonwebtoken": "^9.0.0",
+    "path": "^0.12.7"
```
Comment (vc-mohit) [inline JWT_Auth/package.json]: why it's required to install in package

File: JWT_Auth/src/controllers/authController.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+export const loginUser = async (req, res, next) => {
+  try {
+    const { email, password } = req.body;
+    const dbData = await helper.getFileData();
+    const user = dbData.find((user) => user.email === email);
+    if (!user) return res.status(404).send({ message: "Email not found" });
+    const isPasswordMatch = await bcrypt.compare(password, user.password);
+    if (!isPasswordMatch)
+      return helper.handleResponseSend(res, "Password does not match", 401);
```
Comment (vc-mohit) [inline JWT_Auth/src/controllers/authController.js:14]: check the message

File: JWT_Auth/package.json (hunk 9 above)
Comment (vc-bhanudatsinh) [reply JWT_Auth/package.json]: ok, fixed it

**Comment by vc-bhanudatsinh** [comment]
I have made the changes as said in the CR please review it @vc-mohit and @karansoni-vc

**Review by vc-mohit — APPROVED** [review]

**Pull request merged**